kotlinx-coroutines-core/common/src/channels/Channel.kt

```diff
@@ -242,6 +243,19 @@ public interface ReceiveChannel<out E> {
      */
     public operator fun iterator(): ChannelIterator<E>
 
+    /**
+     * Consumes all elements from this channels by [emitting][FlowCollector.emit] them to the given [collector]
+     * and [cancels][cancel] the channel after the execution of the block.
+     * If you need to iterate over the channel without consuming it, a regular `for` loop should be used instead.
+     *
+     * This function provides a more efficient shorthand for `consumeEach { value -> collector.emit(value) }`.
+     * See [consumeEach].
+     *
+     * @suppress **This an internal API and should not be used from general code.**
+     */
+    @InternalCoroutinesApi
+    public suspend fun consumeEachTo(collector: FlowCollector<E>)
```

Comment:
I wonder why we need such intrusive method for `@FlowPreview` API.
We can write `cosumeAsFlow` on top of an iterator with nulling-out the slot to workaround #1333.
I realize that such an approach is a bit slower, but at least it does not require to add a new internal method to already overloaded channels

Reply:
For a fix to #1333 it can be indeed just an extension. However, converting to/from channels seems to be important part of a flow API surface and writing it in such a way offers better performance as it avoids iterator altogether. Also, the whole state machine and spilling around iterator-using code is much more complicated and the resulting code would be even more kludgy. 

However, there are other way to do the fix. An alternative but also quite performant way would be to introduce `ReceiveOrClosed` (even as experimental/iternal API that would be good here), and then `consumeEachTo` can be written as extension in the flow package. Can we resurrect PR #762 ?

Reply:
I would selfishly love this alternative, I've got one or two other use cases for `ReceiveOrClosed` that I'm working around awkwardly right now.

Reply:
Pulled in PR #762 into here and implemented it in a less intrusive way.